From dbfeca97b4ad3a591d4e8fb095f6c2b6019577b4 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 19 Feb 2024 09:55:43 +0100
Subject: [PATCH 0627/1695] core: io: fix IO_READ32_POLL_TIMEOUT() when delay
 is 0us

Fix detection of timeout condition in IO_READ32_POLL_TIMEOUT() that was
never triggered when delay argument is 0us. Indeed 0 is not a useful
increment value for a timeout counter.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 4edfd4c8cb076128357bf92cc0018cdeeca791d0 core: io: fix IO_READ32_POLL_TIMEOUT() when delay is 0us]
Fixes: 97ea199a2ae8 ("core: io: IO_READ32_POLL_TIMEOUT()")
Acked-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Acked-by: Jerome Forissier <jerome.forissier@linaro.org>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
(cherry picked from commit 98d105a565cef6cc3e861184b7e2a745a985cf49)
Change-Id: Iaf2857965d34d60bfaa1dd07e5fbd0bd32edbcd3
---
 core/include/io.h | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/core/include/io.h b/core/include/io.h
index 0ef91907b..2b4717661 100644
--- a/core/include/io.h
+++ b/core/include/io.h
@@ -7,6 +7,7 @@
 
 #include <compiler.h>
 #include <kernel/delay.h>
+#include <kernel/delay_arch.h>
 #include <stdint.h>
 #include <types_ext.h>
 #include <utee_defines.h>
@@ -287,14 +288,13 @@ static inline void io_clrsetbits8(vaddr_t addr, uint8_t clear_mask,
  */
 #define IO_READ32_POLL_TIMEOUT(_addr, _val, _cond, _delay_us, _timeout_us) \
 	({ \
-		uint32_t __timeout = 0; \
+		uint64_t __timeout = timeout_init_us(_timeout_us); \
 		uint32_t __delay = (_delay_us); \
 		\
-		while (__timeout < (_timeout_us)) { \
+		while (!timeout_elapsed(__timeout)) { \
 			(_val) = io_read32(_addr); \
 			if (_cond) \
 				break; \
-			__timeout += (__delay); \
 			udelay(__delay); \
 		} \
 		(_val) = io_read32(_addr); \
-- 
2.47.2

